### ISSUE TITLE
Dynamo NULL type handling

### ISSUE THREAD

**Issue by dubiousdavid** [guardian/scanamo#49]
We have lot’s of fields that are wrapped in an Option that are currently being stored in Dynamo as a `NULL` type. Scanamo isn’t able to coerce them to a `None`.

**Comment by philwills** [comment]
I think that if you have a custom `DynamoFormat` for `Option` in scope, that should be able to override the default behaviour to deal with this case. e.g.

```scala
import com.gu.scanamo._
import cats.data.Xor
import com.amazonaws.services.dynamodbv2.model.AttributeValue
import com.gu.scanamo.error.DynamoReadError

implicit def nullOptionFormat[T](implicit f: DynamoFormat[T]) = new DynamoFormat[Option[T]]{
  def read(av: AttributeValue): Xor[DynamoReadError, Option[T]] = {
    Option(av).filter(!_.isNULL).map(f.read(_).map(Some(_)))
       .getOrElse(Xor.right(Option.empty[T]))
  }
  def write(t: Option[T]): AttributeValue = t.map(f.write).getOrElse(new AttributeValue().withNULL(true))
}
```

then 

```scala
scala>DynamoFormat[Option[String]].write(Option.empty[String])
res0: com.amazonaws.services.dynamodbv2.model.AttributeValue = {NULL: true,}
```

Does that deal adequately with your case?

**Comment by philwills** [comment]
There's also probably a case for adding that filter on read to the default, but I don't particularly like the idea of things not being symmetric by default.

**Comment by travisbrown** [comment]
@philwills You could add that behavior to the `read` side (to accommodate other systems that use explicit `NULL`) but leave `write` the same. Still not symmetric, but it's a less dramatic change.

**Comment by cb372** [comment]
For the record, I had the same problem when upgrading from an ancient version of Scanamo (serves me right for being an early adopter). Pre-0.3, Scanamo used to encode `None` as an explicit `NULL`.

[https://github.com/guardian/scanamo/commit/9520eca2ab64e5d461e2f2433156d6ca3b5d66e2#diff-4cdca785dc406e510fc950615adf3442R180](Here) is the change.

**Comment by philwills** [comment]
Closed by #65
